### PR TITLE
Testing whether executable exists and whether the file is an exe file

### LINKED ISSUE #38: Add support for infinite looping programs
Check if we are in an infinite loop. Maybe use DFS on the frame variables to see if we have visited an identical configuration in the past? and created a loop

### LINKED ISSUE #39: Check if executable file exists
Check if the executable passed to trace actually exists before running it

### PR DESCRIPTION
**Pull request by cboumalh**
I added a few lines that as far as I know get the task done (checking if executable exists and whether the file passes is an executable in the first place. When you're free check if out and let me know!! thanks man

### CODE REVIEW

File: trace.py [hunk 1]
```diff
@@ -195,7 +195,31 @@ def define_val_type(val): # recursive function used to change strings into typed
...
+    print("This file is not an executable!")
+    os.remove("output.txt")
+    exit()
+
+# Testing time it took program to run
+os.popen(f'(time {sys.argv[1]}) &> time.txt')
+time.sleep(5)
+run_time = open('time.txt')
+if not list(run_time.read().splitlines()):
+    print(list(run_time.read().splitlines()))
```
Comment (cboumalh) [inline trace.py]: This line is unnecessary. It's just so I can see the list contents.

File: trace.py [hunk 2]
```diff
@@ -195,7 +195,31 @@ def define_val_type(val): # recursive function used to change strings into typed
 # Load binary passed in argv[1] and check we only have one arg
 if len(sys.argv) < 2:
     print("Please pass in at least one argument (executable you wish to run)!")
+    os.remove("output.txt")
     exit()
+# Test whether this file exists by checking if the path exists
+if not os.path.exists(sys.argv[1]):
+    print("This executable does not exist!")
+    os.remove("output.txt")
```
Comment (David-Simonetti-ND) [inline trace.py]: I think all the error checking statement here should belong in one function

Comment (David-Simonetti-ND) [reply trace.py]: that way we have one function call to like `check_executable_good` or something and it makes it all more clear.

File: trace.py [hunk 3]
```diff
@@ -195,7 +195,31 @@ def define_val_type(val): # recursive function used to change strings into typed
...
+    exit()
+# Test whether the file is an executable
+if not os.access(sys.argv[1], os.X_OK):
+    print("This file is not an executable!")
+    os.remove("output.txt")
+    exit()
+
+# Testing time it took program to run
+os.popen(f'(time {sys.argv[1]}) &> time.txt')
```
Comment (David-Simonetti-ND) [inline trace.py]: This is a good idea!

File: trace.py [hunk 4]
```diff
@@ -195,7 +195,31 @@ def define_val_type(val): # recursive function used to change strings into typed
...
+time.sleep(5)
+run_time = open('time.txt')
+if not list(run_time.read().splitlines()):
+    print(list(run_time.read().splitlines()))
+    print("There may be an infinite loop in the code")
+    os.remove("output.txt")
+    os.remove("time.txt")
+    run_time.close()
```
Comment (David-Simonetti-ND) [inline trace.py]: What happens to the process that you started with os.popen? is it still running in the background after trace.py exits? If this is how it behaves, I would recommend explicitly killing the process before exiting

**Review by David-Simonetti-ND — CHANGES_REQUESTED** [review]
Very nice overall! I left just a few small comments

**Comment by David-Simonetti-ND** [comment]
Also whats up with the automatic check failing?

File: trace.py (hunk 4 above)
Comment (cboumalh) [reply trace.py]: I will look into that!

**Comment by cboumalh** [comment]
> Also whats up with the automatic check failing?

I'm not sure, I was gonna ask about that. I have no idea why it's happening

**Comment by David-Simonetti-ND** [comment]
I just poked around with your branch, and I realized that currently we have to wait the full 10 seconds every time even if the program runs in 0.01 milliseconds. I found this code on stackoverflow
import shlex
from subprocess import Popen, PIPE
from threading import Timer

def run(cmd, timeout_sec):
    proc = Popen(shlex.split(cmd), stdout=PIPE, stderr=PIPE)
    timer = Timer(timeout_sec, proc.kill)
    try:
        timer.start()
        stdout, stderr = proc.communicate()
    finally:
        timer.cancel()

run("sleep 1", 5)  # process ends normally at 1 second
run("sleep 5", 1)  # timeout happens at 1 second
It runs a process with a given timeout.
And I think that the check is failing because you are not properly redirecting the output to a file. Try checking to make sure the time.txt file is being created.

**Comment by cboumalh** [comment]
> I just poked around with your branch, and I realized that currently we have to wait the full 10 seconds every time even if the program runs in 0.01 milliseconds. I found this code on stackoverflow import shlex from subprocess import Popen, PIPE from threading import Timer
> 
> def run(cmd, timeout_sec): proc = Popen(shlex.split(cmd), stdout=PIPE, stderr=PIPE) timer = Timer(timeout_sec, proc.kill) try: timer.start() stdout, stderr = proc.communicate() finally: timer.cancel()
> 
> run("sleep 1", 5) # process ends normally at 1 second run("sleep 5", 1) # timeout happens at 1 second It runs a process with a given timeout. And I think that the check is failing because you are not properly redirecting the output to a file. Try checking to make sure the time.txt file is being created.

I went over the code and made sure I understand how it works. I'll text you later to discuss it a bit. Thanks for the help!!

**Pull request merged**